### PR TITLE
kirkwood: add support for Lenovo ix2-dl

### PR DESCRIPTION
**Pull request by jdwl1o1**
Lenovo ix2-dl is a dual SATA NAS powered by a Marvell
 Kirkwood SoC clocked at 1.6GHz. It has 256MB of RAM and 1GiB of
 flash memory, 1x USB 2.0 and 1x 1Gbit/s NIC.
 Was also sold by Iomega as the Iomega StorCentre ix2-dl.
 This device is very similar to the ix2-200 however it lacks 2 USB ports and includes
 a larger nand flash unit. The IX2-DL was sold without drives, with the OS stored in nand.

Specification:
- SoC: Marvell Kirkwood 88F6281
- CPU/Speed: 1600Mhz
- Flash-Chip: Samsung NAND
- Flash size: nand: 1024 MiB, SLC, erase size: 128 KiB, page size: 2048, OOB size: 64
- RAM: 256MB
- LAN: 1x 1000 Mbps Ethernet
- WiFi: none
- 1x USB 2.0
- UART1: for serial console. UART0 not available.

Installation instructions (modified version of the ix2-200 commit):
This runs OpenWrt from ram to allow sysupgrade to install to nand.

1. download initramfs-uImage image and sysupgrade image, copy into tftp server or ext2 formatted usb drive
2. access uboot environment with serial cable and run
```
    setenv mainlineLinux yes
    setenv arcNumber 1682
    setenv console 'console=ttyS0,115200'
    setenv mtdparts 'mtdparts=orion_nand:0x100000@0x000000(u-boot)ro,0x20000@0xA0000(u-boot environment)ro,0x300000@0x100000(kernel),0x1C00000@0x400000(ubi)'
    setenv
    setenv bootargs_root 'root='
    setenv bootcmd 'setenv bootargs ${console} ${mtdparts} ${bootargs_root}; nand read.e 0x800000 0x100000 0x300000; bootm 0x00800000'
    saveenv
```
For USB Boot:  
``` 
 usb reset; ext2load usb 0:1 0x00800000 /initramfs.bin; bootm 0x00800000
```
For TFTP boot:
```
    setenv serverip [tftp server ip]    
    setenv ipaddr 192.168.1.13
    tftpboot 0x00800000 factory.bin
    bootm 0x00800000
```
3. ssh to openwrt and sysupgrade to install into flash
```
    mkdir /mnt/usb
    mount /dev/sda1 /mnt/usb
    cp /mnt/usb/*sysupgrade* /tmp
    or load from tftp using tftpboot
    sysupgrade -n /tmp/sysupgrade.bin
```
4. access openwrt by dhcp ip address assigned by your router or at 192.168.1.1

note 1 - sata drives should not be installed when installing
note 2 - this is a 1 way process. Once the nand is overwritten returning to stock will not be possible

Signed-off-by: Jeremy Lobascher <jeremy@lobascher.com>

Thanks for your contribution to OpenWrt!

To help keep the codebase consistent and readable,
and to help people review your contribution,
we ask you to follow the rules you find in the wiki at this link
https://openwrt.org/submitting-patches

Please remove this message before posting the pull request.

### CODE REVIEW

File: target/linux/kirkwood/base-files/etc/board.d/02_network [hunk 1]
```diff
@@ -35,6 +35,9 @@ linksys,ea4500)
 	ucidef_set_interfaces_lan_wan "ethernet1 ethernet2 ethernet3 ethernet4" "internet"
 	ucidef_set_interface_macaddr "wan" $( mtd_get_mac_ascii u_env eth1addr )
 	;;
+"lenovo,ix2")
```
Comment (CHKDSK88) [inline target/linux/kirkwood/base-files/etc/board.d/02_network:38]: duplicate section

File: target/linux/kirkwood/base-files/etc/diag.sh [hunk 2]
```diff
@@ -0,0 +1,58 @@
+#!/bin/sh
```
Comment (CHKDSK88) [inline target/linux/kirkwood/base-files/etc/diag.sh:1]: Please drop it. Status led are configured in dts. Please see an example:
https://github.com/openwrt/openwrt/blob/ee1a92834da4dab409dc4cd150bb188c6b069020/target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-ea3500.dts#L26

File: target/linux/kirkwood/base-files/etc/init.d/hwmon_fancontrol [hunk 3]
```diff
@@ -44,5 +44,18 @@ boot() {
 		# pre-defined steps: 103000, 122000, 143300, 170000 in [millicelsius]
 		echo 103000 > "$path_to_hwmon/temp3_auto_temp_max"
 		;;
+	lenovo,ix2)
```
Comment (CHKDSK88) [inline target/linux/kirkwood/base-files/etc/init.d/hwmon_fancontrol:47]: alphabetical order missed

File: target/linux/kirkwood/patches-5.10/113-lenovo-ix2.patch [hunk 4]
```diff
@@ -0,0 +1,708 @@
+--- a/arch/arm/boot/dts/kirkwood-lenovo_ix2.dts
++++ b/arch/arm/boot/dts/kirkwood-lenovo_ix2.dts
```
Comment (CHKDSK88) [inline target/linux/kirkwood/patches-5.10/113-lenovo-ix2.patch:2]: Please change patch to regular file and move it to: target/linux/kirkwood/files/arch/arm/boot/dts/

File: target/linux/kirkwood/patches-5.10/113-lenovo-ix2.patch [hunk 5]
```diff
@@ -0,0 +1,708 @@
...
+@@ -0,0 +1,705 @@
++/dts-v1/;
++
++/ {
++	#address-cells = < 0x01 >;
++	#size-cells = < 0x01 >;
++	compatible = "lenovo,ix2\0marvell,kirkwood-88f6282\0marvell,kirkwood";
```
Comment (CHKDSK88) [inline target/linux/kirkwood/patches-5.10/113-lenovo-ix2.patch:9]: Something goes wrong here.

File: target/linux/kirkwood/patches-5.10/113-lenovo-ix2.patch [hunk 6]
```diff
@@ -0,0 +1,708 @@
+--- a/arch/arm/boot/dts/kirkwood-lenovo_ix2.dts
++++ b/arch/arm/boot/dts/kirkwood-lenovo_ix2.dts
+@@ -0,0 +1,705 @@
++/dts-v1/;
++
```
Comment (CHKDSK88) [inline target/linux/kirkwood/patches-5.10/113-lenovo-ix2.patch:5]: Please rewrite dts, and use:
```
#include "kirkwood.dtsi"
#include "kirkwood-6281.dtsi"
```

**Review by CHKDSK88 — CHANGES_REQUESTED** [review]
Dts need a huge refactoring. It looks like decompiled from dtb. Some chamges in other file are required too.

**Comment by jdwl1o1** [comment]
Hi,

Thank you for the feedback. 

Unfortunately re-writing the DTS is beyond my capability and would require many more tests to confirm functionality.

**Comment by adschm** [comment]
https://openwrt.org/docs/guide-developer/frequent-pr-mistakes#do_not_multi-post_or_re-post_prs

File: target/linux/kirkwood/base-files/etc/diag.sh (hunk 2 above)
Comment (jdwl1o1) [reply target/linux/kirkwood/base-files/etc/diag.sh]: Removed

File: target/linux/kirkwood/base-files/etc/init.d/hwmon_fancontrol (hunk 3 above)
Comment (jdwl1o1) [reply target/linux/kirkwood/base-files/etc/init.d/hwmon_fancontrol]: Set to Alphabetical Order

File: target/linux/kirkwood/patches-5.10/113-lenovo-ix2.patch (hunk 4 above)
Comment (jdwl1o1) [reply target/linux/kirkwood/patches-5.10/113-lenovo-ix2.patch]: Done, new DTS created

File: target/linux/kirkwood/patches-5.10/113-lenovo-ix2.patch (hunk 5 above)
Comment (jdwl1o1) [reply target/linux/kirkwood/patches-5.10/113-lenovo-ix2.patch]: Patch removed

File: target/linux/kirkwood/patches-5.10/113-lenovo-ix2.patch (hunk 6 above)
Comment (jdwl1o1) [reply target/linux/kirkwood/patches-5.10/113-lenovo-ix2.patch]: Done, new DTS created and tested.

**Comment by jdwl1o1** [comment]
Hi OpenWRT team, I've updated the PR with the corrected files and new DTS. 

I am not very familiar with git - other commits have been picked up that I have not been involved with?

**Comment by jdwl1o1** [comment]
Hi, I have reworked the dts completely, using details from other IX2-DL user' dts files. I've also identified that there are multiple models of 'IX2' from Iomega and Lenovo and so have made this PR specific to the IX2-DL only. 
The 'non-nand' model would be similar and the dts is similar, however I do not have access to one of those to test.

**Comment by adschm** [comment]
What you need is "git rebase" and "git rebase -i", so you put your changes on top of current master, and then squash all "your" changes into a single commit. "git merge" almost never helps for anything (good).

File: target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-lenovo-ix2-dl.dts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+ */
+
+/dts-v1/;
+
+#include "kirkwood-lenovo-ix2-common.dtsi"
+
+/ {
+	model = "Lenovo IX2-dl";
+	compatible = "lenovo-ix2-dl", "marvell,kirkwood-88f6282", "marvell,kirkwood";
+
```
Comment (adschm) [inline target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-lenovo-ix2-dl.dts:20]: remove empty line here.

File: target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-lenovo-ix2-common.dtsi [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+ *
+ * Copyright (c) 2015-2016 bodhi <mibodhi@gmail.com>
+ * Copyright (c) 2015      balbes150
+ *
+ * This file is licensed under the terms of the GNU General Public
+ * License version 2.  This program is licensed "as is" without any
+ * warranty of any kind, whether express or implied.
+ */
+
+/dts-v1/;
```
Comment (adschm) [inline target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-lenovo-ix2-common.dtsi:12]: dts-v1 is only needed once. Please look where others put it (kirdwood.dtsi, other dtsi, or dts).

File: target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-lenovo-ix2-common.dtsi [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+
+	gpio-leds {
+		compatible = "gpio-leds";
+		pinctrl-0 = <&pmx_led_sys_white &pmx_led_sys_red
+			     &pmx_led_hdd_blue
+			     &pmx_led_hdd1_red &pmx_led_hdd2_red>;
+		pinctrl-names = "default";
+
+		white-sys {
+			label = "ix2:white:sys";
```
Comment (adschm) [inline target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-lenovo-ix2-common.dtsi:143]: please remove model prefix, just "white:sys", "red:sys", etc.

Why "sys" and not "system"?

File: target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-lenovo-ix2-common.dtsi [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+			regulator-always-on;
+			regulator-boot-on;
+			enable-active-high;
+			gpio = <&gpio0 17 GPIO_ACTIVE_HIGH>;
+		};
+	};
+};
+
+&mdio {
+	status = "okay";
```
Comment (adschm) [inline target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-lenovo-ix2-common.dtsi:210]: Please add empty line after status (everywhere).

File: target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-lenovo-ix2-dl.dts [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+&nand {
+	status = "okay";
+	chip-delay = <35>;
+
+	partition@0 {
+		label = "uboot";
+		reg = <0x0000000 0x0080000>;
+		read-only;
+	};
```
Comment (adschm) [inline target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-lenovo-ix2-dl.dts:31]: Please add empty lines between nodes.

File: target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-lenovo-ix2-dl.dts [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+		label = "uboot_env2";
+		reg = <0x00c0000 0x0020000>;
+	};
+	partition@100000 {
+		label = "kernel";
+		reg = <0x0100000 0x0300000>;
+	};
+	partition@400000 {
+		label = "ubi";
+		reg = <0x0400000 0x020000000>;
```
Comment (adschm) [inline target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-lenovo-ix2-dl.dts:46]: 512 MiB != 1 GiB

File: target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-lenovo-ix2-dl.dts [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+};
+
+&nand {
+	status = "okay";
+	chip-delay = <35>;
+
+	partition@0 {
+		label = "uboot";
+		reg = <0x0000000 0x0080000>;
```
Comment (adschm) [inline target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-lenovo-ix2-dl.dts:29]: Please use the same number of digits everywhere or remove all leading zeros.

First offset should be just 0x0

File: target/linux/kirkwood/image/Makefile [hunk 14]
```diff
@@ -100,6 +100,22 @@ define Device/iom_ix2-200
 endef
 TARGET_DEVICES += iom_ix2-200
 
+define Device/lenovo-ix2-dl
```
Comment (adschm) [inline target/linux/kirkwood/image/Makefile:103]: levono_ix2-dl

File: target/linux/kirkwood/image/Makefile [hunk 15]
```diff
@@ -100,6 +100,22 @@ define Device/iom_ix2-200
...
+  DEVICE_TITLE := Lenovo IX2-dl
```
Comment (adschm) [inline target/linux/kirkwood/image/Makefile:104]: DEVICE_TITLE has been replaced by DEVICE_VENDOR/DEVICE_MODEL more than a year ago.

File: target/linux/kirkwood/image/Makefile [hunk 16]
```diff
@@ -100,6 +100,22 @@ define Device/iom_ix2-200
...
+  DEVICE_PACKAGES += kmod-gpio-button-hotplug kmod-i2c-mv64xxx kmod-hwmon-adt7475
+  DEVICE_TYPE:=nas
+  PAGESIZE := 512
+  SUBPAGESIZE := 256
+  BLOCKSIZE := 16KiB
+  KERNEL_SIZE := 3072k
+  KERNEL_IN_UBI :=
+  UBINIZE_OPTS := -E 5
+  IMAGE_SIZE := 32505856
+  IMAGE/factory.bin := append-kernel | pad-to $$$$(KERNEL_SIZE) | append-ubi | check-size $$$$(IMAGE_SIZE)
```
Comment (adschm) [inline target/linux/kirkwood/image/Makefile:115]: 74 chars. per line max. (everywhere).
check-size without argument

File: target/linux/kirkwood/image/Makefile [hunk 17]
```diff
@@ -100,6 +100,22 @@ define Device/iom_ix2-200
 endef
 TARGET_DEVICES += iom_ix2-200
 
+define Device/lenovo-ix2-dl
+  DEVICE_TITLE := Lenovo IX2-dl
+  DEVICE_DTS := kirkwood-lenovo-ix2-dl
+  DEVICE_PACKAGES += kmod-gpio-button-hotplug kmod-i2c-mv64xxx kmod-hwmon-adt7475
+  DEVICE_TYPE:=nas
```
Comment (adschm) [inline target/linux/kirkwood/image/Makefile:107]: DEVICE_TYPE cannot be set per device. Drop it.

File: target/linux/kirkwood/image/Makefile [hunk 18]
```diff
@@ -100,6 +100,22 @@ define Device/iom_ix2-200
...
+  PAGESIZE := 512
+  SUBPAGESIZE := 256
+  BLOCKSIZE := 16KiB
+  KERNEL_SIZE := 3072k
+  KERNEL_IN_UBI :=
+  UBINIZE_OPTS := -E 5
+  IMAGE_SIZE := 32505856
```
Comment (adschm) [inline target/linux/kirkwood/image/Makefile:114]: Where does this value come from? Why not use "k"?

File: target/linux/kirkwood/base-files/etc/board.d/02_network [hunk 19]
```diff
@@ -35,6 +30,14 @@ linksys,ea4500)
...
 	ucidef_set_interface_macaddr "wan" $( mtd_get_mac_ascii u_env eth1addr )
 	;;
+raidsonic,ib-nas62x0|\
+seagate,blackarmor-nas220|\
+seagate,dockstar|\
+seagate,goflexhome|\
+seagate,goflexnet|\
+zyxel,nsa310s)
+	ucidef_set_interface_lan "eth0" "dhcp"
+	;;
```
Comment (adschm) [inline target/linux/kirkwood/base-files/etc/board.d/02_network:40]: I do not understand this change.

File: target/linux/kirkwood/base-files/etc/board.d/01_leds [hunk 20]
```diff
@@ -12,6 +12,9 @@ case "$board" in
 iom,ix2-200)
 	ucidef_set_led_timer "health" "health" "status:white:rebuild_led" "200" "800"
 	;;
+lenovo-ix2-dl)
+	ucidef_set_led_timer "power" "power" "status:white:rebuild_led" "200" "800"
+        ;;
```
Comment (adschm) [inline target/linux/kirkwood/base-files/etc/board.d/01_leds:17]: broken indent.

Using ucidef_set_led_timer is discouraged for new devices. Please use led-* aliases instead like all newer devices.

**Review by adschm — CHANGES_REQUESTED** [review]
This seems to be based on a very old sample. Please have a look at devices in the current master to improve your submission.

**Comment by jdwl1o1** [comment]
Hi OpenWRT team, my tree appears to be corrupted or incorrect (or I'm doing something wrong), with every update from master then sending the commits to this PR. As I'm not able to resolve this I am going to close the PR and delete my fork.
My sincere apologies for wasting your time.

**Pull request closed without merging**